### PR TITLE
[Vue] migrate piwik-field and related directives

### PR DESCRIPTION
**Pull request by diosmosis**
### Description:

**This PR is based off of https://github.com/matomo-org/matomo/pull/18292.**

`piwik-field` is a recursive directive so they all had to be converted at once, so this PR is also quite large.

Changes:
- Migrate `piwik-field` directive to the vue Field directive. Note: the `templateFile` property is not longer supported. If plugin developers use it in the wild, we'll have to provide a different way of extending Field.
- Migrate `matomo-multi-pair-field` directive to MultiPairField.
- Migrate `matomo-field-array` directive to FieldArray.
- Add extra parameters to createAngularJsAdapter() scope property transform function.
- Migrate `piwik-form-field` directive to FormField. Every template that was included in the angularjs directive was converted to a new vue component used by FormField.
- Add new mechanism to support extending form fields that allows passing in a Vue component, rather than angularjs template (this method is still supported for BC, until Matomo 5).
- Do not use Vue UMDs if there is a dist folder w/ files but no src folder (can happen during development).
- Fix bug visible in 4.x-dev w/ report export: changing the report limit selector did not change the filter_limit in the export URL. For some reason, angularjs couldn't modify $parent.reportLimit, possibly because there is also a scope.reportLimit.
- Put the defineAsyncComponent() function used to get components from other plugins when we don't know if they're loaded into the new `useExternalPluginComponent()` function.
- Removed the selectedSite internal state in the SiteSelector component, it now follows the v-model contract properly.

image

image

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: core/AssetManager/UIAssetFetcher/JScriptUIAssetFetcher.php
Comment (diosmosis) [inline core/AssetManager/UIAssetFetcher/JScriptUIAssetFetcher.php:108]: During development it can be possible for there to be a dist/ folder but no src/ folder (happens when checking out different branches where one branch doesn't have a vue folder I believe).

File: plugins/CoreHome/CoreHome.php
Comment (diosmosis) [inline plugins/CoreHome/CoreHome.php]: Found myself having to debug angularjs quite often w/ the unminified code, so I thought I'd just add this.

File: plugins/CoreHome/CoreHome.php
Comment (diosmosis) [inline plugins/CoreHome/CoreHome.php:268]: I thought about using an adapter, but since the data in the angularjs service is meant to be mutable from outside of the service, and the Vue counterpoint is not, I thought it would be better to simply leave it for now.

File: plugins/CoreHome/angularjs/report-export/reportexport.directive.js
Comment (diosmosis) [inline plugins/CoreHome/angularjs/report-export/reportexport.directive.js:259]: Using an object was the only way I could get this to work. This bug (described in the issue description) is also in 4.x-dev.

File: plugins/CoreHome/angularjs/report-export/reportexport.popover.html
Comment (diosmosis) [inline plugins/CoreHome/angularjs/report-export/reportexport.popover.html:41]: Note: `full-width="false"` never seemed to work.

File: plugins/CoreHome/vue/src/SiteSelector/SiteSelector.adapter.ts
Comment (diosmosis) [inline plugins/CoreHome/vue/src/SiteSelector/SiteSelector.adapter.ts]: sitename is assumed to be encoded in places.

File: plugins/CoreHome/vue/src/createAngularJsAdapter.ts
Comment (diosmosis) [inline plugins/CoreHome/vue/src/createAngularJsAdapter.ts:84]: AngularJS sometimes adds its own properties to scope properties (like `$$hashKey`), this removes them when sending them to Vue.

File: plugins/CoreHome/vue/src/createAngularJsAdapter.ts
Comment (diosmosis) [inline plugins/CoreHome/vue/src/createAngularJsAdapter.ts:159]: Seemed to work w/o kebab case, but as far as I know it shouldn't, so changed it.

File: plugins/CoreHome/vue/src/createAngularJsAdapter.ts
Comment (diosmosis) [inline plugins/CoreHome/vue/src/createAngularJsAdapter.ts:317]: Utility functions to deal w/ `@` angularjs attributes that are meant to be booleans/numbers.

File: plugins/CorePluginsAdmin/vue/src/Field/Field.adapter.ts
Comment (diosmosis) [inline plugins/CorePluginsAdmin/vue/src/Field/Field.adapter.ts:166]: Converts the angularjs `component` property which will be a JSON object to a Vue component. From Vue, we'll just pass a component directly.

File: tests/lib/screenshot-testing/support/page-renderer.js
Comment (diosmosis) [inline tests/lib/screenshot-testing/support/page-renderer.js:127]: waitFor was not working before since it didn't return the promises returned by these functions. It would return immediately instead of waiting.

File: tests/lib/screenshot-testing/support/page-renderer.js
Comment (diosmosis) [inline tests/lib/screenshot-testing/support/page-renderer.js]: puppeteer types too fast for vue to update the model, so adding a timeout here fixed a bunch of UI tests.

File: CHANGELOG.md
Comment (sgiehl) [inline CHANGELOG.md]:
```suggestion
## Matomo 4.7.0
```

File: plugins/CoreHome/CoreHome.php
Comment (sgiehl) [reply plugins/CoreHome/CoreHome.php]: In theory that might cause problems if someone enables development mode in a Matomo not installed from git. As the the unminified files is removed in builds.

File: tests/lib/screenshot-testing/support/page-renderer.js
Comment (sgiehl) [reply tests/lib/screenshot-testing/support/page-renderer.js:127]: good catch. Guess I didn't notice as I directly converted all occurrences 🙈

**Comment by sgiehl** [comment]
@diosmosis I'm currently doing a review. Might take a while as the changes are quite huge.
Looking through some of the expected UI file changes I have already seen, that the title for text `FieldArray`s are not displayed for each input anymore. See https://github.com/matomo-org/matomo/commit/d343d20a429bc86b205fb9762c498df84a566896?short_path=ecaef42#diff-ecaef42b8254004961e2206553ed6b910ac38155a6fa329264534e400d068fc0
Feel free to fix that already, while I'm still reviewing the code.

File: plugins/CorePluginsAdmin/vue/src/FormField/FormField.vue
Comment (sgiehl) [inline plugins/CorePluginsAdmin/vue/src/FormField/FormField.vue]:
```suggestion
        // for checkboxes, the value MUST be either true or false
```

File: plugins/CorePluginsAdmin/vue/src/FormField/FormField.vue
Comment (sgiehl) [inline plugins/CorePluginsAdmin/vue/src/FormField/FormField.vue]: Can this comment be removed or is anything left over?

File: plugins/CoreHome/angularjs/report-export/reportexport.popover.html
Comment (sgiehl) [inline plugins/CoreHome/angularjs/report-export/reportexport.popover.html]: Unrelated to this issue. But I guess this actually should be `maxFilterLimit`.

**Review by sgiehl — CHANGES_REQUESTED** [review]
Left a couple of comments while looking through the code only. Will now check tests & do some local testing.

Btw. We won't be able to merge this one, until #18406 is merged, as your branch already contains some of the commits from `next_release` branch.

**Comment by sgiehl** [comment]
@diosmosis While clicking through the UI / tests I so far spotted this issues:

- [x] Opening the configuration for GoogleAnalyticsImporter (which is configured) results in a JavaScript error:
  image
  The same also happens for CustomTranslations plugin, same error caused by `multitupletext` field
- [x] The expected test file for mobile messaging changed here: https://github.com/matomo-org/matomo/commit/f0d1f279d552105955150e4442366a751937c9af?short_path=0ad0702#diff-0ad0702aa329c2b750bd3f088bc8316ac7d524d1b98f06a2e01d930dfd4d53ed
   I don't think we should update the test file, but adjust the test instead to work the same again. The problem seems to be, that you kind of fixed an issue in select boxes. Before the mobile provider select box had an empty entry in the beginning. After the vue changes that empty entry is gone (which imho is correct). So guess we need to adjust this: https://github.com/matomo-org/matomo/blob/9db3daa0349a2a08a8a50b28537722a611c3868d/tests/UI/specs/UIIntegration_spec.js#L818
- [x] Opening the AB-Tests management, results in: 
   image
   Same happens for FormAnalytics and when editing a goal
- [x] Opening the tracking code generator doesn't show the tracking codes anymore till some property is changed. I think this regressed earlier, as it worked in 4.5.0, but doesn't on 4.x-dev. (That's also the reason for the [failing tests](https://app.travis-ci.com/github/matomo-org/matomo/jobs/549742225#L1679)) If that's unrelated to Vue, feel free to move that to a new issue.
- [x] On the GDRP Tools page, there is another error:
   image

**Comment by sgiehl** [comment]
I'm done for now. Couldn't find anything else while clicking around. Feel free to re-request a review once everything is fixed.

**Comment by diosmosis** [comment]
@sgiehl this should be ready for review again

**Review by sgiehl — APPROVED** [review]
@diosmosis I have clicked through the issues I had reported and everything seems to work fine now. One last thing needs to be updated. It seems you have updated the CustomAlerts submodule to a specific branch, this might cause problems when we merge that. If there are changes required in the submodule please create PRs there. Guess we need to reset the submodule to `4.x-dev` here before merging. Otherwise good to merge, if tests are passing again.

**Comment by diosmosis** [comment]
@sgiehl I'll create a PR on CustomAlerts :+1:

**Comment by sgiehl** [comment]
fixed a failing test. Will merge this now. @diosmosis feel free to ping me, once the PR for custom alerts is created.

**Pull request merged**
